src/cell.cpp

```diff
@@ -635,6 +635,143 @@ void DAGCell::to_hdf5(hid_t group_id) const { return; }
 
 #endif
 
+//==============================================================================
+// UniversePartitioner implementation
+//==============================================================================
+
+UniversePartitioner::UniversePartitioner(const Universe& univ)
+{
+  // Find all of the z-planes in this universe.  Add them to the surfs_ member.
+  // Use surf_set for O(1) searches that ensure surfs_ entries are not repeated.
+  std::unordered_set<int32_t> surf_set;
+  for (auto i_cell : univ.cells_) {
+    for (auto token : model::cells[i_cell]->rpn_) {
+      if (token < OP_UNION) {
+        auto i_surf = std::abs(token) - 1;
+        const auto* surf = model::surfaces[i_surf].get();
+        if (const auto* zplane = dynamic_cast<const SurfaceZPlane*>(surf)) {
+          if (surf_set.find(i_surf) == surf_set.end()) {
+            surf_set.insert(i_surf);
+            surfs_.push_back(i_surf);
+          }
+        }
+      }
+    }
+  }
+
+  // Define a lambda for comparing z-planes by their location, and use it to
+  // sort the surfs_ member.
+  std::sort(surfs_.begin(), surfs_.end(),
+    [](int32_t i_surf, int32_t j_surf) -> bool
+    {
+      const auto* surf = model::surfaces[i_surf].get();
+      const auto* zplane = dynamic_cast<const SurfaceZPlane*>(surf);
+      double zi = zplane->z0_;
+      surf = model::surfaces[j_surf].get();
+      zplane = dynamic_cast<const SurfaceZPlane*>(surf);
+      double zj = zplane->z0_;
+      return zi < zj;
+    }
+  );
+
+  // Populate the partition lists.
+  partitions_.resize(surfs_.size() + 1);
+  for (auto i_cell : univ.cells_) {
+    // Find the tokens for bounding z-planes.
+    int32_t lower_token = 0, upper_token = 0;
+    double min_z, max_z;
+    for (auto token : model::cells[i_cell]->rpn_) {
+      if (token < OP_UNION) {
+        const auto* surf = model::surfaces[std::abs(token) - 1].get();
+        if (const auto* zplane = dynamic_cast<const SurfaceZPlane*>(surf)) {
+          if (lower_token == 0 || zplane->z0_ < min_z) {
+            lower_token = token;
+            min_z = zplane->z0_;
+          }
+          if (upper_token == 0 || zplane->z0_ > max_z) {
+            upper_token = token;
+            max_z = zplane->z0_;
+          }
+        }
+      }
+    }
```

Comment:
This block of code could be combined with the first block where you are building `surf_set` since it's all the same up until the point you're comparing to `min_z`.

Reply:
Nevermind, I take that back. I see this is specific for each cell.